File: mobile/src/main/java/com/bclymer/dailybudget/models/BudgetStats.java
package com.bclymer.dailybudget.models;

import com.bclymer.dailybudget.utilities.Util;

import java.util.Calendar;
import java.util.Comparator;
import java.util.Date;
import java.util.GregorianCalendar;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

/**
 * Created by brianclymer on 9/7/15.
 */
public class BudgetStats {

    private static final String TAG = BudgetStats.class.getSimpleName();

    public double totalSpent;
    public double totalAllowance;
    public long totalDays;
    public Map<String, Double> places = new HashMap<>();

    private Budget budget;

    public BudgetStats(Budget budget) {
        this.budget = budget;

        List<Transaction> transactions = budget.getSortedTransactions();
        if (transactions.size() == 0) {
            Util.toast("No Transactions");
            return;
        }

        Transaction firstTransaction = transactions.get(transactions.size() - 1);
        Calendar firstDate = new GregorianCalendar();
        firstDate.setTime(firstTransaction.date);
        Util.setCalendarToBeginningOfDay(firstDate);
        totalDays = Util.getDaysBetweenDates(firstTransaction.date, new Date()) + 1;

        for (Transaction transaction : transactions) {
            if (transaction.location == null || transaction.location.trim().length() == 0 || transaction.getTotalAmount() > 0) {
                if (transaction.getTotalAmount() > 0) {
                    totalAllowance += transaction.getTotalAmount();
                }
                continue;
            }

            totalSpent -= transaction.getTotalAmount();
            if (places.containsKey(transaction.location)) {
                places.put(transaction.location, places.get(transaction.location) - transaction.getTotalAmount());
            } else {
                places.put(transaction.location, -transaction.getTotalAmount());
            }
        }

        places = sortByValue(places);
    }

    public double getSpentPerDay() {
        return totalSpent / totalDays;
    }

    public double getCurrentAllowance() {
        return budget.cachedValue;
    }

    public Map<String, Double> sortByValue(Map<String, Double> unsortedMap) {
        Map<String, Double> sortedMap = new TreeMap<>(new ValueComparator(unsortedMap));
        sortedMap.putAll(unsortedMap);
        return sortedMap;
    }

    static class ValueComparator implements Comparator<String> {

        final Map<String, Double> map;

        public ValueComparator(Map<String, Double> map) {
            this.map = map;
        }

        public int compare(String keyA, String keyB) {
            Double valueA = map.get(keyA);
            Comparable<Double> valueB = map.get(keyB);
            return valueB.compareTo(valueA);
        }
    }
}

File: mobile/src/main/kotlin/com/bclymer/dailybudget/models/BudgetStats.kt
package com.bclymer.dailybudget.models

import com.bclymer.dailybudget.utilities.Util
import java.util.*

/**
 * Created by brianclymer on 9/7/15.
 */
class BudgetStats(private val budget: Budget, private val transactions: List<Transaction>) {

    var totalSpent: Double = 0.toDouble()
    var totalAllowance: Double = 0.toDouble()
    var totalDays: Long = 0
    var places: MutableMap<String, Double> = HashMap()

    init {
        if (transactions.size == 0) {
            Util.toast("No Transactions")
        } else {
            val firstTransaction = transactions[transactions.size - 1]
            val firstDate = GregorianCalendar()
            firstDate.time = firstTransaction.date
            Util.setCalendarToBeginningOfDay(firstDate)
            totalDays = Util.getDaysBetweenDates(firstTransaction.date, Date()) + 1

            for (transaction in transactions) {
                if (transaction.totalAmount > 0) {
                    totalAllowance += transaction.totalAmount
                    continue
                }

                val location: String
                if (transaction.location == null || transaction.location.trim { it <= ' ' }.length == 0) {
                    location = "<None>"
                } else {
                    location = transaction.location
                }

                totalSpent -= transaction.totalAmount
                if (places.containsKey(location)) {
                    places.put(location, places[location]!! - transaction.totalAmount)
                } else {
                    places.put(location, -transaction.totalAmount)
                }
            }

            places = sortByValue(places)
        }
    }

    val spentPerDay: Double
        get() = totalSpent / totalDays

    val currentAllowance: Double
        get() = budget.cachedValue

    fun sortByValue(unsortedMap: Map<String, Double>): MutableMap<String, Double> {
        val sortedMap = TreeMap<String, Double>(ValueComparator(unsortedMap))
        sortedMap.putAll(unsortedMap)
        return sortedMap
    }

    internal class ValueComparator(val map: Map<String, Double>) : Comparator<String> {

        override fun compare(keyA: String, keyB: String): Int {
            val valueA = map[keyA]!!
            val valueB = map[keyB]!!
            return valueB.compareTo(valueA)
        }
    }

    companion object {
        private val TAG = BudgetStats::class.java.simpleName
    }
}
